miden-lib/asm/faucets/basic.masm

```diff
@@ -0,0 +1,91 @@
+# BASIC FUNGIBLE FAUCET CONTRACT
+# =================================================================================================
+# This is a basic fungible faucet smart contract. 
+#
+# It allows the owner of the faucet to mint, distribute, and burn tokens. Token metadata is stored 
+# at position 1 as [token_symbol, decimals, 0, max_supply], where:
+# - token_symbol is expressed as <TODO add explainer>.
+# - decimals are the decimals of the token.
+# - max_supply is the maximum supply of the token.
+use.miden::sat::account
+use.miden::sat::asset
+use.miden::sat::faucet
+use.miden::sat::tx
+use.miden::eoa::basic->basic_eoa
+
+# CONSTANTS
+# =================================================================================================
+#
+# Slot in memory to store the metadata
+const.METADATA_SLOT=1
+
+export.basic_eoa::auth_tx_rpo_falcon512
+
+#! Distributes freshly minted fungible assets to the provided address.
+#! 
+#! Inputs: [amount, tag, RECIPIENT, ...]
+#! Outputs: [note_ptr, 0, 0, 0, 0, 0, 0, 0, 0, ...]
+#!
+#! - amount is the amount to be minted and sent.
+#! - tag is the tag to be included in the note.
+#! - RECIPIENT is the recipient of the asset, i.e., 
+#!   hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash).
+#! - note_ptr is the pointer to the memory address in the kernel.
+#!   This cannot directly be accessed from another context.
+#!
+#! FAILS if: 
+#! - The transaction is being executed against an account that is not a fungible asset faucet.
+#! - The total issuance after minting is greater than the maximum allowed supply.
+export.distribute
+    # get max supply of this faucet. We assume it is stored at pos 3 of slot 1
+    push.METADATA_SLOT exec.account::get_item drop drop drop 
+    # => [max_supply, amount, tag, RECIPIENT, ...]
+    
+    # get total issuance of this faucet so far and add amount to be minted
+    exec.faucet::get_total_issuance
+    # => [total_issuance, max_supply, amount, tag, RECIPIENT, ...]
+
+    # compute maximum amount that can be minted, max_mint_amount = max_supply - total_issuance
+    sub
+    # => [max_supply - total_issuance, amount, tag, RECIPIENT, ...]
+
+    # check that amount =< max_supply - total_issuance, fails if otherwise
+    dup.1 swap lte assert
+    # => [asset, tag, RECIPIENT, ...]
+
+    # creating the asset
+    exec.asset::create_fungible_asset
+    # => [ASSET, tag, RECIPIENT, ...]
+    
+    # minting the asset (note: asset preservation logic)
```

Comment:
minor nit: maybe clarify the statement in parenthesis. For example: "mint the asset; this is needed to satisfy asset preservation logic".